src/coreclr/jit/importer.cpp

```diff
@@ -967,21 +972,28 @@ GenTree* Compiler::impStoreStruct(GenTree*         store,
             GenTreeFlags indirFlags = GTF_EMPTY;
             GenTree*     destAddr   = impGetNodeAddr(store, CHECK_SPILL_ALL, &indirFlags);
 
+            // Make sure we don't pass something other than a local address to the return buffer arg.
+            // It is allowed to pass current's method return buffer as it is a local too.
             if (!impIsAddressInLocal(destAddr) &&
-                (!destAddr->OperIsScalarLocal() || (destAddr->AsLclVarCommon()->GetLclNum() != info.compRetBuffArg)))
+                (!destAddr->OperIsScalarLocal() ||
+                 (destAddr->AsLclVarCommon()->GetLclNum() != impInlineRoot()->info.compRetBuffArg)))
             {
                 unsigned tmp = lvaGrabTemp(false DEBUGARG("stack copy for value returned via return buffer"));
                 lvaSetStruct(tmp, call->gtRetClsHnd, false);
                 destAddr = gtNewLclVarAddrNode(tmp, TYP_BYREF);
+
                 // Insert address of temp into existing call
-                call->gtArgs.InsertAfterThisOrFirst(this,
-                                                    NewCallArg::Primitive(destAddr).WellKnown(WellKnownArg::RetBuffer));
+                NewCallArg retBufArg = NewCallArg::Primitive(destAddr).WellKnown(WellKnownArg::RetBuffer);
+                call->gtArgs.InsertAfterThisOrFirst(this, retBufArg);
+
                 // Now the store needs to copy from the new temp instead.
                 store->Data() = gtNewLclvNode(tmp, lvaGetDesc(tmp)->TypeGet());
```

Comment:
I think it would be better to add the comma here, so that you call `impStoreStruct` on the store with the comma as the data. Then `impStoreStruct` will remove the embedded statement automatically in favor of appending (when possible)